engine/runtime-parser/src/test/java/org/enso/compiler/core/PersistanceTest.java

```diff
@@ -0,0 +1,327 @@
+    assertSame("Head and tail are the same", out.head(), out.last());
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void scalaListSharedRefAtDepth() throws Exception {
+    var idLoc1 = Singleton.INSTANCE;
+    var in = join(Option.apply(idLoc1), join(Option.apply(idLoc1), nil()));
+
+    var out = serde(List.class, in, -1);
+
+    assertEquals("Two elements", 2, out.size());
+    var readHead = (Option<Singleton>) out.head();
+    var readTail = (Option<Singleton>) out.last();
+
+    assertEquals("Head is equal to original", idLoc1, readHead.get());
+    assertEquals("Tail is equal to original", idLoc1, readTail.get());
+
+    assertNotSame("Head and tail are different", readHead, readTail);
+    assertSame("Head and tail are the same", readHead.get(), readTail.get());
+  }
+
+  @Test
+  public void lazyScalaSequence() throws Exception {
+    var s1 = new LazySeq("Hello");
+    var s2 = new LazySeq("World");
+
+    var second = new boolean[1];
+    @SuppressWarnings("unchecked")
+    var in =
+        (Seq<String>)
+            Seq.fill(
+                2,
+                () -> {
+                  if (second[0]) {
+                    return s2;
+                  }
+                  second[0] = true;
+                  return s1;
+                });
+    assertEquals("Seq with two elements created", 2, in.length());
+
+    LazySeq.forbidden = true;
+    var out = serde(Seq.class, in, -1);
+
+    assertEquals("Two elements", 2, out.size());
+
+    LazySeq.forbidden = false;
+
+    assertEquals("Lazily deserialized s2", s2, out.head());
+    assertNotSame("Lazily deserialized s2", s2, out.head());
+    assertEquals("Lazily deserialized s1", s1, out.last());
+    assertNotSame("Lazily deserialized s1", s1, out.head());
+  }
+
+  @Test
+  public void serializeModule() throws Exception {
+    var meta = new MetadataStorage(nil());
+    var diag = new DiagnosticStorage(nil());
+    var m = new Module(nil(), nil(), nil(), true, Option.empty(), meta, diag);
+
+    var out = serde(Module.class, m, -1);
+
+    assertEquals("Same", m, out);
+  }
+
+  @Test
+  public void hashMap() throws Exception {
+    var map = new HashMap<String, String>();
+    map.put("one", "uno");
+    map.put("two", "duo");
+    map.put("ten", "tre");
+
+    var out = serde(HashMap.class, map, -1);
+
+    assertEquals("Same", map, out);
+  }
+
+  @Test
+  public void withService() throws Exception {
+    var in = new Service(5);
+    var arr = Persistance.writeObject(in);
+
+    var plain = Persistance.readObject(arr);
+    assertEquals("Remains five", 5, plain.get(Service.class).value());
+
+    var multi = Persistance.readObject(arr, new Service(3));
+    assertEquals("Multiplied", 15, multi.get(Service.class).value());
+  }
+
+  private static <T> T serde(Class<T> clazz, T l, int expectedSize) throws IOException {
+    var arr = Persistance.writeObject(l);
+    if (expectedSize >= 0) {
+      assertEquals(expectedSize, arr.length - 12);
```

Comment:
What is this 12 magic constant here?

Reply:
It is the [size of the header](https://github.com/enso-org/enso/pull/8207/files#r1392573405) before the actual stream is written down.